### DISCUSSION TITLE
RandAffined does not work as expected

### DISCUSSION THREAD

**Discussion by Spenhouet** [Project-MONAI/MONAI#3319] [Q&A]
**Describe the bug**

Given the following image and label:
![image](https://user-images.githubusercontent.com/7819068/141314133-b37f21c1-48eb-413c-bba4-21c4e774f745.png)

When applying `RandAffined`, the image and labels no longer match:

![image](https://user-images.githubusercontent.com/7819068/141314287-5711ee04-cb1e-4cd7-9c55-451fcd0528b5.png)

**To Reproduce**

1. Load these files: [example_files.zip](https://github.com/Project-MONAI/MONAI/files/7521159/example_files.zip)
2. Execute this pipeline:
```python
from pathlib import Path

import numpy as np
from monai.transforms.compose import Compose
from monai.transforms.io.dictionary import LoadImaged, SaveImaged
from monai.transforms.spatial.dictionary import RandAffined
from monai.transforms.utility.dictionary import (EnsureChannelFirstd, ToTensord)
from monai.utils.enums import GridSampleMode

path = Path('scratch/rand_affine')

T1_KEY = 't1'
LABEL_KEY = 'labels'
FILE_KEYS = [T1_KEY, LABEL_KEY]
SAMPLE_MODES = [GridSampleMode.BILINEAR, GridSampleMode.NEAREST]

data = {
    T1_KEY: path / 't1.nii.gz',
    LABEL_KEY: path / 'labels.nii.gz',
}

process = Compose([
    LoadImaged(FILE_KEYS),
    ToTensord(FILE_KEYS),
    EnsureChannelFirstd(FILE_KEYS),
    RandAffined(
        FILE_KEYS,
        rotate_range=5.0 * np.pi / 180.0,  # type: ignore
        prob=1.0,
        mode=SAMPLE_MODES,
        allow_missing_keys=True,
    ),
    SaveImaged(
        FILE_KEYS,
        output_dir=path,
        output_postfix=f"out",
        resample=False,
        output_dtype=np.float32,
        separate_folder=False,
        print_log=False,
    )
])

process([data])
```

**Expected behavior**
The same random affine transformation is applied to all images.
After the transform the labels still match the image.

**Additional context**
I did not yet identify the issue.

**Comment by wyli** [comment]
I couldn't replicate the issue locally with a macbook and visualise with itk-snap.
<img width="776" alt="Screenshot 2021-11-11 at 15 51 27" src="https://user-images.githubusercontent.com/831580/141327964-e01303f6-91a0-43d1-808a-4e0cc6599118.png">

could you help provide more info?

**Comment by Spenhouet** [comment]
@wyli You did not use the data provided by me. Could you try it with that? It is part of the reproduction.

**Comment by wyli** [comment]
> @wyli You did not use the data provided by me. Could you try it with that? It is part of the reproduction.

I downloaded and unzipped the file, there are two files, `t1.nii.gz` and `labels.nii.gz`. itk-snap refuses to display 
<img width="398" alt="Screenshot 2021-11-11 at 15 58 35" src="https://user-images.githubusercontent.com/831580/141328988-a1856c6e-2dce-4071-930e-00eb5e071bc7.png">
the labels.nii.gz as segmentations for t1.nii.gz

**Reply by Spenhouet** [reply]
@wyli Why did you change this to a discussion? 
You just used the wrong tool to visualize images. That is not the fault of this bug report and also does not change anything with respect to the bug.

Please reopen the bug report.

**Comment by Spenhouet** [comment]
I'm not a itk-snap user but if there is a problem in itk-snap probably good to report it to their project. Not really the issue I'm reporting here though.

Try using MRIcroGL or Slicer. Both work fine.

**Comment by Spenhouet** [comment]
@Nic-Ma @ericspod @wyli @rijobro This was closed due to an unrelated issue with itk-snap. I don't know why this was the reason for closing this bug report? Could you clarify and reopen the issue?

**Reply by rijobro** [reply]
Is it possible that what we're seeing here is due to differences in image and label sizes? I haven't checked the `RandAffined` code, but wouldn't be surprised if that's what's causing it. This is similar to your `CropForegroundd` problem, I don't know if there's a plan to support differing image/label sizes in MONAI as I suspect it would be a large undertaking.

**Reply by Spenhouet** [reply]
@rijobro Yes, it is. I reopened the issue and clarified that part: https://github.com/Project-MONAI/MONAI/issues/3339

With respect to the "if", I really think this should be supported. Honestly, to me the "same image size" case is a rather toy dataset case. So to me that is honestly not a real question. I strongly believe that MONAI should support this.

**Reply by rijobro** [reply]
I respectfully disagree that DL use cases with matching image/label sizes are limited to toy examples, as in my experience they represent the overwhelming majority of cases that we use. I also think that your problem could be solved by simply cropping your larger images to match your smaller ones, or padding your smaller ones to match your larger ones. Seems like a fairly easy initial transform that would solve all these problems you're seeing.

**Reply by Spenhouet** [reply]
As I replied here https://github.com/Project-MONAI/MONAI/issues/3182#issuecomment-968757425, it is not the case that there is an easy work around.
If it would help and you are interested, we could get on a call and I could show you our use case. This would clarify on why this is necessary.

> I respectfully disagree that DL use cases with matching image/label sizes are limited to toy examples, as in my experience they represent the overwhelming majority of cases that we use.

My statement might have been badly worded. I do agree that the majority of use cases will have matching images. I was absolutely not stating that these are all toy examples. I meant to say that there are much more use cases beyond that. Like, if we deviate from the norm, we can no longer use MONAI?
